src/main/java/org/commcare/cases/query/QueryContext.java

```diff
@@ -0,0 +1,141 @@
+package org.commcare.cases.query;
+
+import org.commcare.cases.query.queryset.CurrentModelQuerySet;
+import org.commcare.cases.query.queryset.QuerySetCache;
+import org.javarosa.core.model.condition.EvaluationContext;
+import org.javarosa.core.model.trace.EvaluationTrace;
+
+/**
+ * A Query Context object is responsible for keeping track of relevant metadata about where a
+ * query is executing that may make it possible for the planner to better identify when to
+ * trigger certain query handlers.
+ *
+ * For instance, if an individual query is only looking for one matching case, it may do a single
+ * DB read. The context can provide a cue that the query is likely to be run over many other cases,
+ * which can provide the planner with a hint to fetch those cases in bulk proactively
+ *
+ * The QueryContext Object's lifecycle is also used to limit the scope of any of that bulk caching.
+ * Since the object lifecycle is paired with the EC of the query, large chunks of memory can be
+ * allocated into this object and it will be removed when the context is no longer relevant.
+ *
+ * Created by ctsims on 1/26/2017.
+ */
+
+public class QueryContext {
+
+    private static final int BULK_QUERY_THRESHOLD = 50;
+
+    //TODO: This is a bad reason to keep the EC around here, and locks the lifecycle of this object
+    //into the EC
+    private EvaluationContext traceRoot;
+
+    private QueryCacheHost cache;
+
+    private QueryContext potentialSpawnedContext;
+
+    //Until we can keep track more robustly of the individual spheres of 'bulk' models
+    //we'll just keep track of the dominant factor in our queries to know what to expect
+    //WRT whether optimizatons will help or hurt
+    private int contextScope = 1;
+
+    public QueryContext() {
+        cache = new QueryCacheHost();
+    }
+
+    private QueryContext(QueryContext parent) {
+        this.traceRoot = parent.traceRoot;
+        this.cache = new QueryCacheHost(parent.cache);
+        this.contextScope = parent.contextScope;
+    }
+
+    /**
+     * @param newScope the magnitude of the new query
+     * @return either the existing QueryContext or a new (child) QueryContext
+     * if the magnitude of the new query exceeds the parent sufficiently
+     */
+    public QueryContext checkForDerivativeContextAndReturn(int newScope) {
+        QueryContext newContext;
+
+        //TODO: I think we may need to clear the spanwed context's spawned context (maybe?) if it
```

Comment:
I'm not sure if comments like this are helpful, we have a bunch of them all over the codebase that never get addressed because no one knows what to do with them. I would be in favor of either deciding this is important enough to address now, or removing this.